staging/src/k8s.io/client-go/tools/pager/pager.go

```diff
@@ -118,9 +118,12 @@ func (p *ListPager) List(ctx context.Context, options metav1.ListOptions) (runti
 		// initialize the list and fill its contents
 		if list == nil {
 			list = &metainternalversion.List{Items: make([]runtime.Object, 0, options.Limit+1)}
-			list.ResourceVersion = m.GetResourceVersion()
 			list.SelfLink = m.GetSelfLink()
 		}
+
+		// the resource version of the last page is the resource version for the list
+		list.ResourceVersion = m.GetResourceVersion()
```

Comment:
doesn't this break anyone starting a watch from the list RV (like informers do) if an item from an earlier page changed or was deleted before a later page's resource version?

Reply:
The resource version of any page returned in a list has to be useable to watch consistently.  Think of each page having more info about the list, but if an inconsistent RV is returned, that’s a bug anyway.  The advantage of “last page RV” is it allows some consistency to be resolved late and leaves the door open for longer term optimizations.  An example is a key space where a-j is served by one etcd and k-z is served by another - you can do a list consistently across a-j with only the RV from etcd1, and once you start reading k your RV has to be a composite of etcd1 and etcd2 (for watch).  But if you stop paging before you hit etcd2, the RV for watch doesn’t  have to be a composite.

Reply:
> Think of each page having more info about the list, but if an inconsistent RV is returned, that’s a bug anyway. 

We literally have a conformance test which verifies that we can return an inconsistent RV and continue to page off of it. https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/test/e2e/apimachinery/chunking.go#L124

Reply:
> We literally have a conformance test which verifies that we can return an inconsistent RV and continue to page off of it.

That's different... that is only useful for a client that wants to observe the full key space and doesn't care if different portions of the key space were at different versions. The inconsistent RV is communicated as a field in a status error (because for most clients, it is an error condition).

Reply:
> The resource version of any page returned in a list has to be useable to watch consistently. Think of each page having more info about the list, but if an inconsistent RV is returned, that’s a bug anyway. The advantage of “last page RV” is it allows some consistency to be resolved late and leaves the door open for longer term optimizations.

hmm... so this is relaxing a client-side restriction to enable a potential future server modification which would return different-but-consistent RVs for different pages?

is there a server-side test that ensures paging returns the same RV for all pages so this relaxation doesn't open the door for accidental breakage in the current implementation?

Reply:
> The inconsistent RV is communicated as a field in a status error (because for most clients, it is an error condition).

Yeah I realize how it's used, but I'm not a fan of it. People accidentally ignore errors all the time and in this case, it will cause really terrible things to possibly happen. 

I would prefer we just expose a startKey so that if you want to observe the entire keyspace without consistency, you can do so explicitly.

Reply:
This isn’t about running without consistency.  But let’s separate the scenarios.

No matter what, we have to define what the RV returned by a page is (currently it’s implemented as identical across all pages, but not specified as such).  The continue token allows you to continue a paging request consistently.  The RV allows you to restart a list at a roughly approximate time *or* start a consistent watch.  We don’t actually guarantee any longer you can restart a consistent watch by any mechanism because of the change to list RV (for watch cache) to mean “newer than”.   That’s important because the continue token is now a stronger guarantee than list resourceVersion (although watch resourceVersion is as strong as continue).

I’m not opposed to startKey (as part of the kep to address inconsistent key space scanning which is valuable for lots of client types), but it’s not directly relevant to the guarantees provided between list paging and watch establishment.